src/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -0,0 +1,221 @@
+--
+-- Shape: if the list of runs is empty and the optional input tree is
+-- structurally empty, the result will also be structurally empty. See
+-- 'isStructurallyEmpty'.
+--
+-- Resource tracking:
+-- * This allocates a new 'Ref' which the caller is responsible for releasing
+--   eventually.
+-- * The ownership of all input 'Ref's remains with the caller. This action
+--   will create duplicate references, not adopt the given ones.
+--
+-- ASYNC: this should be called with asynchronous exceptions masked because it
+-- allocates\/creates resources.
+newPendingLevelMerge ::
+     forall m h.
+     (MonadMVar m, MonadMask m, PrimMonad m)
+  => [PreExistingRun m h]
+  -> Maybe (Ref (MergingTree m h))
+  -> m (Ref (MergingTree m h))
+newPendingLevelMerge [] (Just t) = dupRef t
+newPendingLevelMerge prs mmt = do
+    -- There are no interruption points here, and thus provided async exceptions
+    -- are masked then there can be no async exceptions here at all.
+    mergeTreeState <- case (prs, mmt) of
+      ([PreExistingRun r], Nothing) ->
+        CompletedTreeMerge <$> dupRef r
+
+      ([PreExistingMergingRun mr], Nothing) ->
+        OngoingTreeMerge <$> dupRef mr
+
+      _ -> PendingTreeMerge <$>
+            (PendingLevelMerge <$> traverse dupPreExistingRun prs
+                               <*> dupMaybeMergingTree mmt)
+
+    newMergeTree mergeTreeState
+  where
+    dupPreExistingRun (PreExistingRun r) =
+      PreExistingRun <$> dupRef r
+    dupPreExistingRun (PreExistingMergingRun mr) =
+      PreExistingMergingRun <$> dupRef mr
+
+    dupMaybeMergingTree :: Maybe (Ref (MergingTree m h))
+                        -> m (Maybe (Ref (MergingTree m h)))
+    dupMaybeMergingTree Nothing   = return Nothing
+    dupMaybeMergingTree (Just mt) = do
+      isempty <- isStructurallyEmpty mt
+      if isempty
+        then return Nothing
+        else Just <$> dupRef mt
+
+-- | Create a new 'MergingTree' representing the union of one or more merging
+-- trees. This is for unioning the content of multiple tables (represented
+-- themselves as merging trees).
+--
+-- Shape: if all of the input trees are structurally empty, the result will
+-- also be structurally empty. See 'isStructurallyEmpty'.
+--
+-- Resource tracking:
+-- * This allocates a new 'Ref' which the caller is responsible for releasing
+--   eventually.
+-- * The ownership of all input 'Ref's remains with the caller. This action
+--   will create duplicate references, not adopt the given ones.
+--
+-- ASYNC: this should be called with asynchronous exceptions masked because it
+-- allocates\/creates resources.
+newPendingUnionMerge ::
+     (MonadMVar m, MonadMask m, PrimMonad m)
+  => [Ref (MergingTree m h)]
+  -> m (Ref (MergingTree m h))
+newPendingUnionMerge [mt] = dupRef mt
+newPendingUnionMerge mts = do
+    mts' <- mapM dupRef =<< filterM (fmap not . isStructurallyEmpty) mts
+    newMergeTree (PendingTreeMerge (PendingUnionMerge mts'))
```

Comment:
We should probably first filter, then pattern match on `[mt]`. If there are multiple inputs, but only one is non-empty, we don't need to create a new tree.

Reply:
Yes fair enough. What I had is fine for correctness, but isn't optimal.